### PR TITLE
8327093: Add truncate function to BitMap API

### PR DESCRIPTION
**Pull request by matias9927**
The BitMap does not currently have a "slice" method which can return a subset of the bitmap defined by a start and end bit. This utility can be used to analyze portions of a bitmap or to overwrite the existing map with a shorter, modified map. 

This implementation has two core methods: 
`copy_of_range` allocates and returns a new word array as used internally by BitMap
`truncate` overwrites the internal array with the one generated by `copy_of_range` so none of the internal workings are exposed to callers outside of BitMap.

This change is verified with an included test.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8327093](https://bugs.openjdk.org/browse/JDK-8327093): Add truncate function to BitMap API (**Enhancement** - P4)

### Reviewers
 * [Axel Boldt-Christmas](https://openjdk.org/census#aboldtch) (@xmas92 - **Reviewer**)
 * [Ioi Lam](https://openjdk.org/census#iklam) (@iklam - **Reviewer**) ⚠️ Review applies to [17ac26f2](https://git.openjdk.org/jdk/pull/18092/files/17ac26f237171cd5ba3f6c76e33db71493970176)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18092/head:pull/18092` \
`$ git checkout pull/18092`

Update a local copy of the PR: \
`$ git checkout pull/18092` \
`$ git pull https://git.openjdk.org/jdk.git pull/18092/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18092`

View PR using the GUI difftool: \
`$ git pr show -t 18092`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18092.diff">https://git.openjdk.org/jdk/pull/18092.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18092#issuecomment-1977517004)

### CODE REVIEW

**Review by xmas92 — COMMENTED** [review]
The parameter here `bool clear = true` does nothing and it is unclear to me why it is added. 

That `truncate(n)` is `[n:]` and not `[:n]` does not feel intuitive to me. Maybe having a `truncate_start` as well as `truncate_end` which covers both the explicit cases.

The `copy_of_range` logic seems broken. It is caught in the 32-bit gtests. Maybe it should be extended to cover more edge cases. Maybe add some different striped bitmaps.

**Comment by iklam** [comment]
> That `truncate(n)` is `[n:]` and not `[:n]` does not feel intuitive to me. Maybe having a `truncate_start` as well as `truncate_end` which covers both the explicit cases.

The current treatment of the 1-parameter case is similarly to the Java API of [`String.substring(int start)`](https://docs.oracle.com/javase/8/docs/api/java/lang/String.html#substring-int-), or the JavaScript [`slice()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/slice) API for arrays. So if we want the current behavior, it might be better to rename the function as `subset()` or `slice()`.

However,  `subset()` or `slice()` usually mean making a copy of an existing BitMap. Yet, with the current usage of C++ templates in bitMap.hpp, that's very difficult to make a copy of the BitMap object itself. E.g, many of the BitMaps are currently stack-allocated. It's unclear where a "copy" of such a BitMap object should live. (Note: `BitMapWithAllocator::allocate()` allocates a `bm_word_t` array, not a BitMap object).

I think  the name`truncate()` will align with the existing `resize()` API -- both of them modify the BitMap object in-place.

However, a `truncate(n)` that means `[:n]` would do the exact same thing as `resize(n)`. It would be confusing to have two APIs that do the same thing.

But I agree that a `truncate(n)` that means `[n:]` is also non-intuiative.

So I think the best compromise is simply to not offer a single parameter version of `truncate(n)`. It would be fairly easy for the caller to specify `truncate(0, n)` or `truncate(n, bitmap->size_in_bits())` to indicate exactly what is needed.

File: src/hotspot/share/utilities/bitMap.cpp [hunk 1]
```diff
@@ -99,6 +99,70 @@ void GrowableBitMap<BitMapWithAllocator>::resize(idx_t new_size_in_bits, bool cl
...
+  // All words need to be shifted by this amount
+  idx_t shift = bit_in_word(start_bit);
+  // Bits shifted out by a word need to be passed into the next
+  idx_t carry = 0;
+
+  // Iterate the map backwards as the shift will result in carry-out bits
+  for (idx_t i = end_word; i --> start_word;) {
+    // First iteration is a special case:
+    // There may be left over bits in the last word that we want to keep while discarding the rest
+    if (i == end_word - 1 && cutoff > 0) {
```
[the remark this answers is not in the document]
Comment (matias9927) [reply src/hotspot/share/utilities/bitMap.cpp:126]: I suppose it isn't necessary to clear the leftover MSB so that detail can be excluded. The junk bits will still be present however and that may introduce some confusion when debugging.

**Comment by matias9927** [comment]
> But I agree that a `truncate(n)` that means `[n:]` is also non-intuiative.
Looking back at this, I do agree with this and I think it's best for the user to input the entire range. That being said, I think it should be fine to keep `copy_of_range(n)` to signify `[n:]` since it is acting more inline with functions like `slice`.

 Since  `copy_of_range()` is a private function meant to be used within BitMap, I think either `copy_of_range()`, `subset()`, or `slice()` are appropriate names.

**Comment by iklam** [comment]
> I think it should be fine to keep `copy_of_range(n)` to signify `[n:]` since it is acting more inline with functions like `slice`.

But no one uses `copy_of_range(n)` today, so I think we shouldn't add such a variant until the need actually arises.

File: test/hotspot/gtest/utilities/test_bitMap_truncate.cpp [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+  }
+
+  template <class ResizableBitMapClass>
+  static void testTruncateSame() {
+    // Resulting map should be the same as the original
+    ResourceMark rm;
+    ResizableBitMapClass map(BITMAP_SIZE);
+    fillBitMap(map, BITMAP_SIZE);
+    testTruncate<ResizableBitMapClass>(0, BITMAP_SIZE, map);
+  }
```
Comment (iklam) [inline test/hotspot/gtest/utilities/test_bitMap_truncate.cpp:83]: `testTruncateOneWord()` is pretty easy to understand, but I found it really hard to follow the logic of the other test cases. For example, it took me a while to understand the variable `map` in `testTruncateSame()` is not the same as the `map` in `testTruncate()`.

Could all the tests be rewritten to the same style of `testTruncateOneWord()`, where the entire logic is inside the same function?

Comment (iklam) [reply test/hotspot/gtest/utilities/test_bitMap_truncate.cpp:83]: Also, I think the contents of this file should be moved inside test_bitMap.cpp, so you don't need to repeat things like TestArenaBitMap.

File: src/hotspot/share/utilities/bitMap.cpp [hunk 3]
```diff
@@ -99,6 +99,59 @@ void GrowableBitMap<BitMapWithAllocator>::resize(idx_t new_size_in_bits, bool cl
...
+  idx_t const start_word = to_words_align_down(start_bit);
+  idx_t const end_word = to_words_align_up(end_bit);
+  bm_word_t* const old_map = map();
+
+  BitMapWithAllocator* derived = static_cast<BitMapWithAllocator*>(this);
+
+  bm_word_t* new_map = derived->allocate(end_word - start_word);
+
+  // All words need to be shifted by this amount
+  idx_t shift = bit_in_word(start_bit);
```
Comment (iklam) [inline src/hotspot/share/utilities/bitMap.cpp:118]: `shift` should be const as well.

**Review by iklam — CHANGES_REQUESTED** [review]

**Review by iklam — APPROVED** [review]
LGTM

**Comment by iklam** [comment]
> LGTM

BTW, the title of the PR should be changed to drop the word "slice", as it's no used in the code at all. Also, we are not returning a copy of a bitmap, but rather truncating the contents of the current bitmap.

File: src/hotspot/share/utilities/bitMap.cpp [hunk 4]
```diff
@@ -99,6 +99,54 @@ void GrowableBitMap<BitMapWithAllocator>::resize(idx_t new_size_in_bits, bool cl
...
 
+template <class BitMapWithAllocator>
+bm_word_t* GrowableBitMap<BitMapWithAllocator>::copy_of_range(idx_t start_bit, idx_t end_bit) {
+  assert(start_bit < end_bit, "End bit must come after start bit");
+  assert(end_bit <= size(), "End bit not in bitmap");
+
+  // We might have extra bits at the end that we don't want to lose
+  idx_t const cutoff = bit_in_word(end_bit);
+  idx_t const start_word = to_words_align_down(start_bit);
+  idx_t const end_word = to_words_align_up(end_bit);
```
Comment (xmas92) [inline src/hotspot/share/utilities/bitMap.cpp:110]: Using right CV-qualifiers is very rare in HotSpot. I think there are around 600 uses and 1/3 is in the G1 code. 

I do not mind either style, but this PR mixes them, and this file already uses left CV-qualifiers. It would probably be best to keep this consistent.

File: src/hotspot/share/utilities/bitMap.cpp [hunk 5]
```diff
@@ -99,6 +99,54 @@ void GrowableBitMap<BitMapWithAllocator>::resize(idx_t new_size_in_bits, bool cl
...
+  bm_word_t* const old_map = map();
```
Comment (xmas92) [inline src/hotspot/share/utilities/bitMap.cpp:111]: The pointed to type should probably also be `const`. This code should not touch the old memory.

File: src/hotspot/share/utilities/bitMap.cpp [hunk 6]
```diff
@@ -99,6 +99,54 @@ void GrowableBitMap<BitMapWithAllocator>::resize(idx_t new_size_in_bits, bool cl
...
+
+  BitMapWithAllocator* derived = static_cast<BitMapWithAllocator*>(this);
```
Comment (xmas92) [inline src/hotspot/share/utilities/bitMap.cpp:113]: Variable could be `const` (just like `old_map` above)

File: src/hotspot/share/utilities/bitMap.cpp [hunk 7]
```diff
@@ -99,6 +99,54 @@ void GrowableBitMap<BitMapWithAllocator>::resize(idx_t new_size_in_bits, bool cl
...
+
+  bm_word_t* new_map = derived->allocate(end_word - start_word);
```
Comment (xmas92) [inline src/hotspot/share/utilities/bitMap.cpp:115]: Variable could be `const`

File: src/hotspot/share/utilities/bitMap.cpp [hunk 8]
```diff
@@ -99,6 +99,54 @@ void GrowableBitMap<BitMapWithAllocator>::resize(idx_t new_size_in_bits, bool cl
...
+
+  // All words need to be shifted by this amount
+  idx_t const shift = bit_in_word(start_bit);
+  // Bits shifted out by a word need to be passed into the next
+  idx_t carry = 0;
```
Comment (xmas92) [inline src/hotspot/share/utilities/bitMap.cpp:120]: `carry` should be of type `bm_word_t` it contains bits of the payload, not an index.

File: src/hotspot/share/utilities/bitMap.cpp [hunk 9]
```diff
@@ -99,6 +99,54 @@ void GrowableBitMap<BitMapWithAllocator>::resize(idx_t new_size_in_bits, bool cl
...
+  return new_map;
+}
+
+template <class BitMapWithAllocator>
+void GrowableBitMap<BitMapWithAllocator>::truncate(idx_t start_bit, idx_t end_bit) {
+  const size_t old_size_in_words = calc_size_in_words(size());
+  const idx_t new_size_in_bits = end_bit - start_bit;
+  bm_word_t* const old_map = map();
+
+  bm_word_t* new_map = copy_of_range(start_bit, end_bit);
```
Comment (xmas92) [inline src/hotspot/share/utilities/bitMap.cpp:141]: Variable could be `const`

File: src/hotspot/share/utilities/bitMap.cpp [hunk 10]
```diff
@@ -99,6 +99,54 @@ void GrowableBitMap<BitMapWithAllocator>::resize(idx_t new_size_in_bits, bool cl
...
+
+  BitMapWithAllocator* derived = static_cast<BitMapWithAllocator*>(this);
```
Comment (xmas92) [inline src/hotspot/share/utilities/bitMap.cpp:143]: Variable could be `const`

**Review by xmas92 — CHANGES_REQUESTED** [review]
The implementation looks correct now. But there is a mix of styles with regards to const that I think should be consistent at least inside the PR. It is not really possible to unify completely with the rest of the file as there is a mix of `const` usage.

The actual style is not that important to me, but mixing styles with `const` makes me question the intent of code.

I would also have much more confidence in the implementation (and for capturing regressions) if this test could be added.
```diff
diff --git a/test/hotspot/gtest/utilities/test_bitMap.cpp b/test/hotspot/gtest/utilities/test_bitMap.cpp
index e1fc744f07a..2dcb5cfef26 100644
--- a/test/hotspot/gtest/utilities/test_bitMap.cpp
+++ b/test/hotspot/gtest/utilities/test_bitMap.cpp
@@ -148,6 +148,38 @@ class BitMapTruncateTest {
     EXPECT_TRUE(map.is_same(result));
   }
 
+  template <class ResizableBitMapClass>
+  static void testRandom() {
+    for (int i = 0; i < 100; i++) {
+      ResourceMark rm;
+
+      const size_t max_size = 1024;
+      const size_t size = os::random() % max_size + 1;
+      const size_t truncate_size = os::random() % size + 1;
+      const size_t truncate_start = size == truncate_size ? 0 : os::random() % (size - truncate_size);
+
+      ResizableBitMapClass map(size);
+      ResizableBitMapClass result(truncate_size);
+
+      for (BitMap::idx_t idx = 0; idx < truncate_start; idx++) {
+        if (os::random() % 2 == 0) {
+          map.set_bit(idx);
+        }
+      }
+
+      for (BitMap::idx_t idx = 0; idx < truncate_size; idx++) {
+        if (os::random() % 2 == 0) {
+          map.set_bit(truncate_start + idx);
+          result.set_bit(idx);
+        }
+      }
+
+      map.truncate(truncate_start, truncate_start + truncate_size);
+
+      EXPECT_TRUE(map.is_same(result));
+    }
+  }
+
   template <class ResizableBitMapClass>
   static void testTruncateSame() {
     // Resulting map should be the same as the original
@@ -394,3 +426,12 @@ TEST_VM(BitMap, truncate_one_word) {
   BitMapTruncateTest::testTruncateOneWord<TestArenaBitMap>();
   EXPECT_FALSE(HasFailure()) << "Failed on type ArenaBitMap";
 }
+
+TEST_VM(BitMap, truncate_random) {
+  BitMapTruncateTest::testRandom<ResourceBitMap>();
+  EXPECT_FALSE(HasFailure()) << "Failed on type ResourceBitMap";
+  BitMapTruncateTest::testRandom<TestCHeapBitMap>();
+  EXPECT_FALSE(HasFailure()) << "Failed on type CHeapBitMap";
+  BitMapTruncateTest::testRandom<TestArenaBitMap>();
+  EXPECT_FALSE(HasFailure()) << "Failed on type ArenaBitMap";
+}

```

Also agree with @iklam that the PR title should not mention `slice` but `truncate`

**Review by xmas92 — APPROVED** [review]
lgtm.

**Comment by matias9927** [comment]
Thanks for the reviews @iklam and @xmas92!
/integrate
